src/routes/getUpdates.ts

```diff
@@ -0,0 +1,148 @@
+import { asNumber, asObject, asString } from 'cleaners'
+import Router from 'express-promise-router'
+
+import { dataStore } from '../db'
+import {
+  asStoreDirectoryDocument,
+  asStoreFileDocument,
+  asStoreRepoDocument,
+  StoreDirectoryDocument,
+  StoreFileDocument,
+  StoreFileMap,
+  StoreRepoDocument
+} from '../types'
+import { getNameFromPath, makeApiClientError } from '../utils'
+
+export const getUpdatesRouter = Router()
+
+type GetUpdatesBody = ReturnType<typeof asGetUpdatesBody>
+const asGetUpdatesBody = asObject({
+  repoId: asString,
+  timestamp: asNumber
+})
+
+interface GetUpdatesResponseData {
+  paths: StoreFileMap
+  deleted: StoreFileMap
+}
+
+interface DirectoryLike {
+  paths: StoreFileMap
+  deleted: StoreFileMap
+}
+
+getUpdatesRouter.post('/getUpdates', async (req, res) => {
+  let body: GetUpdatesBody
+  let repoId: string
+  let timestamp: number
+
+  try {
+    body = asGetUpdatesBody(req.body)
+    repoId = body.repoId
+    timestamp = body.timestamp
+  } catch (error) {
+    throw makeApiClientError(400, error.message)
+  }
+
+  const repoKey = `${repoId}:/`
+  const repoDocumentResult = await dataStore.get(repoKey)
+
+  const responseData: GetUpdatesResponseData = {
+    paths: {},
+    deleted: {}
+  }
+
+  let repoDocument: StoreRepoDocument
+
+  try {
+    repoDocument = asStoreRepoDocument(repoDocumentResult)
+  } catch (err) {
+    throw makeApiClientError(422, `'${repoKey}' is not a repo`)
+  }
+
+  const { paths, deleted } = await getDirectoryUpdates(
+    repoKey.slice(0, -1),
+    repoDocument,
+    timestamp
+  )
+
+  responseData.paths = paths
+  responseData.deleted = deleted
+
+  res.status(200).json(responseData)
+})
+
+async function getDirectoryUpdates(
+  dirKey: string,
+  dir: DirectoryLike,
+  timestamp: number
+): Promise<DirectoryLike> {
+  const rtn: DirectoryLike = {
+    paths: {},
+    deleted: {}
+  }
+
+  const pathsKeys = Object.keys(dir.paths).map(path => [dirKey, path].join('/'))
+  const deletedKeys = Object.keys(dir.deleted).map(path =>
+    [dirKey, path].join('/')
+  )
+
+  const keysMapOfProp = {
+    paths: pathsKeys,
+    deleted: deletedKeys
+  }
+
+  for (const prop of Object.keys(keysMapOfProp)) {
+    const keys = keysMapOfProp[prop]
+    if (keys.length > 0) {
+      const results = await dataStore.fetch({ keys })
```

Comment:
No need to fetch all of the documents at this point. Really you should check the timestamp first